consensus/checks.go

```diff
@@ -166,3 +184,70 @@ func (consensus *Consensus) onPreparedSanityChecks(
 
 	return true
 }
+
+func (consensus *Consensus) viewChangeSanityCheck(msg *msg_pb.Message) bool {
+	senderKey, err := consensus.verifyViewChangeSenderKey(msg)
+	if err != nil {
+		consensus.getLogger().Error().Err(err).Msgf(
+			"[%s] VerifySenderKey Failed",
+			msg.GetType().String(),
+		)
+		return false
+	}
+	if err := verifyMessageSig(senderKey, msg); err != nil {
+		consensus.getLogger().Error().Err(err).Msgf(
+			"[%s] Failed To Verify Sender's Signature",
+			msg.GetType().String(),
+		)
+		return false
+	}
+	return true
+}
+
+func (consensus *Consensus) onViewChangeSanityCheck(recvMsg *FBFTMessage) bool {
+	// TODO: if difference is only one, new leader can still propose the same committed block to avoid another view change
+	// TODO: new leader catchup without ignore view change message
+	if consensus.blockNum > recvMsg.BlockNum {
+		consensus.getLogger().Debug().
+			Uint64("MsgBlockNum", recvMsg.BlockNum).
+			Msg("[onViewChange] Message BlockNum Is Low")
+		return false
+	}
+	if consensus.blockNum < recvMsg.BlockNum {
+		consensus.getLogger().Warn().
+			Uint64("MsgBlockNum", recvMsg.BlockNum).
+			Msg("[onViewChange] New Leader Has Lower Blocknum")
+		return false
+	}
+	if consensus.current.Mode() == ViewChanging &&
+		consensus.current.ViewID() > recvMsg.ViewID {
+		consensus.getLogger().Warn().
+			Uint64("MyViewChangingID", consensus.current.ViewID()).
+			Uint64("MsgViewChangingID", recvMsg.ViewID).
+			Msg("[onViewChange] ViewChanging ID Is Low")
+		return false
+	}
+	if recvMsg.ViewID-consensus.current.ViewID() > MaxViewIDDiff {
+		consensus.getLogger().Debug().
+			Uint64("MsgViewID", recvMsg.ViewID).
+			Uint64("MaxViewIDDiff", MaxViewIDDiff).
+			Msg("Received viewID that is MaxViewIDDiff further from the current viewID!")
+		return false
+	}
+	return true
+}
+
+func (consensus *Consensus) onNewViewSanityCheck(recvMsg *FBFTMessage) bool {
+	if recvMsg.ViewID < consensus.current.ViewID() {
+		consensus.getLogger().Warn().
+			Uint64("MyViewChangingID", consensus.current.ViewID()).
+			Uint64("MsgViewChangingID", recvMsg.ViewID).
+			Msg("[onNewView] ViewChanging ID Is Low")
+		return false
+	}
+	if !recvMsg.SenderPubkey.IsEqual(consensus.LeaderPubKey) {
```

Comment:
this may not work because for the node that's out of sync with others in ViewID, he won't be able to accept a valid newView msg.

Reply:
we can remove this check I think. As the code below handles the case where the claimed leader (newLeader or some validator) has not achieved quorum. 
```
if !consensus.Decider.IsQuorumAchievedByMask(m3Mask) {
		consensus.getLogger().Warn().
			Msgf("[onNewView] Quorum Not achieved")
		return
	}
```

Reply:
Yea, I actually asked Peckshield to revisit the bug. The solution could actually be to only accept newView when in viewChange mode.

Reply:
removed the senderKey check against consensus.LeaderPubKey check.